Add explain documentation for knn query (#9741)
opensearch-project/documentation-website: _query-dsl/specialized/k-nn/k-nn-explain.md

---
layout: default
title: k-NN query explain
parent: k-NN
grand_parent: Specialized queries
nav_order: 10
---

# k-NN query explain
**Introduced 3.0**
{: .label .label-purple }

You can provide the `explain` parameter to understand how scores are calculated, normalized, and combined in `knn` queries. When enabled, it provides detailed information about the scoring process for each search result. This includes revealing the score normalization techniques used, how different scores were combined, and the calculations for individual subquery scores. This comprehensive insight makes it easier to understand and optimize your `knn` query results. For more information about `explain`, see [Explain API]({{site.url}}{{site.baseurl}}/api-reference/explain/).

`explain` is an expensive operation in terms of both resources and time. For production clusters, we recommend using it sparingly for the purpose of troubleshooting.
{: .warning }

You can provide the `explain` parameter in a URL when running a complete `knn` query for the Faiss engine using the following syntax:

```json
GET <index>/_search?explain=true
POST <index>/_search?explain=true
```

> `explain` for k-NN search for all types of queries with Lucene engine does not return detailed explanation as with the Faiss engine.

Reviewer: ```suggestion
`explain` for k-NN search for all types of queries with the Lucene engine does not return a detailed explanation as with the Faiss engine.
```

{: .note }

The `explain` parameter works for the following types of k-NN search with the Faiss engine:

- [Approximate k-NN search]({{site.url}}{{site.baseurl}}/vector-search/vector-search-techniques/approximate-knn/)
- Approximate k-NN search with [exact search]({{site.url}}{{site.baseurl}}/vector-search/vector-search-techniques/knn-score-script/)
- [Disk-based search]({{site.url}}{{site.baseurl}}/vector-search/optimizing-storage/disk-based-vector-search/)
- [k-NN search with efficient filtering]({{site.url}}{{site.baseurl}}/vector-search/filter-search-knn/efficient-knn-filtering/)
- [Radial search]({{site.url}}{{site.baseurl}}/vector-search/specialized-operations/radial-search-knn/)
- k-NN search with a `term` query

> `explain` for k-NN Search with nested fields does not return detailed explanation like other searches.

Reviewer: ```suggestion
`explain` for k-NN search with nested fields does not return a detailed explanation as with other searches.
```

        "_source" : {
          "my_vector2" : [
            4.5,
            5.5,
            6.7,
            3.7
          ],
          "price" : 4.4
        },
        "_explanation" : {
          "value" : 83.2,
          "description" : "sum of:",
          "details" : [
            {
              "value" : 82.2,
              "description" : "the type of knn search executed was Approximate-NN",
              "details" : [
                {
                  "value" : 82.2,
                  "description" : "the type of knn search executed at leaf was Approximate-NN with vectorDataType = FLOAT, spaceType = innerproduct where score is computed as `-rawScore + 1` from:",
                  "details" : [
                    {
                      "value" : -81.2,
                      "description" : "rawScore, returned from FAISS library",
                      "details" : [ ]
                    }
                  ]
                }
              ]
            },
            {
              "value" : 1.0,
              "description" : "price:[1082969293 TO 1082969293]",
              "details" : [ ]
            }
          ]
        }
      }
    ]
  }
}
```

> </details>

Reviewer 2: ```suggestion

</details>
```
